### ISSUE TITLE
kernel-image-puller is not supporting to pull images from private registries

### ISSUE THREAD

**Issue by savon-noir** [jupyter-server/enterprise_gateway#759]
kernel-image-puller is not supporting to pull images from private registries. I suppose that the best way would be to enable pulling images from private registries by passing KIP_IMAGE_REGISTRY_USER and KIP_IMAGE_REGISTRY_PASSWORD env variables... In a kubernetes setup this could be easily used in combination with a secret containing the user and login.

The python docker client supports this, using the auth_config parameter to docker.images.pull():

https://github.com/docker/docker-py/blob/master/docker/models/images.py

```

    def pull(self, repository, tag=None, **kwargs):
        Args:
            repository (str): The repository to pull
            tag (str): The tag to pull
            auth_config (dict): Override the credentials that are found in the
                config for this request.  ``auth_config`` should contain the
                ``username`` and ``password`` keys to be valid.
            platform (str): Platform in the format ``os[/arch[/variant]]``
```

**Comment by kevin-bates** [comment]
I think this sounds like a great idea!  I'm assuming this could be configured via the helm chart or an equivalent mechanism - correct?  I'm also hoping your intention is provide a PR for this.  :smile:

Thanks for raising this!

**Comment by savon-noir** [comment]
hello Kevin,

indeed my intention was to make a pull req (already started forking it :)) but I thought that maybe i could find a cleaner option for kubernetes which would not require any modification to the kip python code.. and here it is:

The secret:

```
apiVersion: v1
kind: Secret
metadata:
  name: registry-secret
data:
  .dockerconfigjson: <base64 encoded $HOME/.docker/config.json with required creds for the private repo in it (like a service account)>
type: kubernetes.io/dockerconfigjson
```

The daemonset
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kernel-image-puller
  namespace: enterprise-gateway
spec:
  selector:
    matchLabels:
      name: kernel-image-puller
  template:
    metadata:
      labels:
        name: kernel-image-puller
        app: enterprise-gateway
        component: kernel-image-puller
    spec:
      imagePullSecrets:
        - name: registry-secret
      containers:
      - name: kernel-image-puller
        image: my.private.repo.org/elyra/kernel-image-puller:2.0.0
        imagePullPolicy: IfNotPresent
        env:
        - name: KIP_GATEWAY_HOST
          value: "http://enterprise-gateway.enterprise-gateway:8888"
        - name: KIP_INTERVAL
          value: "300"
        - name: KIP_PULL_POLICY
          value: "IfNotPresent"
        volumeMounts:
        - name: dockersock
          mountPath: "/var/run/docker.sock"
        - name: dockerconfig
          mountPath: "/root/.docker"
      volumes:
      - name: dockersock
        hostPath:
          path: /var/run/docker.sock
      - name: dockerconfig
        secret:
          secretName: registry-secret
          items:
          - key: .dockerconfigjson
            path: config.json
```

as simple as this... Of course, this could be directly supported within the helm chart. I can later propose a PR for this if the approach suits you.

Note that I also add to adapt the kernel.json of the different kernel pods template to add support for private registries. 

Thanks a lot for the quick feedback.

R.

**Comment by kevin-bates** [comment]
Thank you for the information - this looks great.  We would just need to make sure that `imagePullSecrets` is conditionally included (when converting to helm).  Looking forward to your PR!

In that vein, I'm going to reopen this issue and we'll ensure there's a "Fixes" clause in the PR that references this.  That way the issue stays tied to the PR.

**Comment by kevin-bates** [comment]
This should be addressed in #1073.  Please see https://github.com/jupyter-server/enterprise_gateway/pull/1053#issuecomment-1098430848.

I'm closing this for now.  Should you find this still unresolved, we can re-open it at that time.
